src/webgpu/idl/javascript.spec.ts

```diff
@@ -0,0 +1,339 @@
+          : desc;
+        return addAllFeatures(adapter, desc);
+      },
+      keyModifier(baseKey: string) {
+        return `${baseKey}:AllFeaturesTest`;
+      },
+    };
+    super.requestDeviceWithRequiredParametersOrSkip(
+      initUncanonicalizedDeviceDescriptor(descriptor),
+      mod
+    );
+  }
+}
+
+/**
+ * A Test that requests all the max limits from the adapter on the device.
+ */
+class AllFeaturesTest extends GPUTest {
+  public static override MakeSharedState(
+    recorder: TestCaseRecorder,
+    params: TestParams
+  ): GPUTestSubcaseBatchState {
+    return new AllFeaturesGPUTestSubcaseBatchState(recorder, params);
+  }
+}
+
+const kResourceInfo = {
+  buffer: {
+    create(t: GPUTest) {
+      return t.createBufferTracked({ size: 16, usage: GPUBufferUsage.UNIFORM });
+    },
+    requiredKeys: [
+      'destroy',
+      'getMappedRange',
+      'label',
+      'mapAsync',
+      'mapState',
+      'size',
+      'unmap',
+      'usage',
+    ],
+  },
+  texture: {
+    create(t: GPUTest) {
+      return t.createTextureTracked({
+        size: [2, 3],
+        format: 'r8unorm',
+        usage: GPUTextureUsage.TEXTURE_BINDING,
+      });
+    },
+    requiredKeys: [
+      'createView',
+      'depthOrArrayLayers',
+      'destroy',
+      'dimension',
+      'format',
+      'height',
+      'label',
+      'mipLevelCount',
+      'sampleCount',
+      'usage',
+      'width',
+    ],
+  },
+  querySet: {
+    create(t: GPUTest) {
+      return t.createQuerySetTracked({
+        type: 'occlusion',
+        count: 2,
+      });
+    },
+    requiredKeys: ['count', 'destroy', 'label', 'type'],
+  },
+  adapter: {
+    create(t: GPUTest) {
+      return t.adapter;
+    },
+    requiredKeys: [
+      'featureLevel',
+      'features',
+      'info',
+      'isCompatibilityMode',
+      'isFallbackAdapter',
+      'limits',
+      'requestDevice',
+    ],
+  },
+  device: {
+    create(t: GPUTest) {
+      return t.device;
+    },
+    requiredKeys: [
+      'adapterInfo',
+      'addEventListener',
+      'createBindGroup',
+      'createBindGroupLayout',
+      'createBuffer',
+      'createCommandEncoder',
+      'createComputePipeline',
+      'createComputePipelineAsync',
+      'createPipelineLayout',
+      'createQuerySet',
+      'createRenderBundleEncoder',
+      'createRenderPipeline',
+      'createRenderPipelineAsync',
+      'createSampler',
+      'createShaderModule',
+      'createTexture',
+      'destroy',
+      'dispatchEvent',
+      'features',
+      'importExternalTexture',
+      'label',
+      'limits',
+      'lost',
+      'onuncapturederror',
+      'popErrorScope',
+      'pushErrorScope',
+      'queue',
+      'removeEventListener',
+    ],
+  },
+  'adapter.limits': {
+    create(t: GPUTest) {
+      return t.adapter.limits;
+    },
+    requiredKeys: kLimits,
+  },
+  'device.limits': {
+    create(t: GPUTest) {
+      return t.device.limits;
+    },
+    requiredKeys: kLimits,
+  },
```

Comment:
Can also test the keys on GPUSupportedFeatures:
https://webidl.spec.whatwg.org/#idl-setlike:~:text=If%20the%20readonly%20keyword%20is%20used%2C%20this%20includes%20entries%2C%20forEach%2C%20has%2C%20keys%2C%20values%2C%20%25Symbol.iterator%25%20methods%2C%20and%20a%20size%20getter.
```
  'adapter.features': {
    create: (t: GPUTest) => t.adapter.features,
    requiredKeys: ['entries', 'forEach', 'has', 'keys', 'values', Symbol.iterator, 'size'],
  },
  'device.features': {
    create: (t: GPUTest) => t.device.features,
    requiredKeys: ['entries', 'forEach', 'has', 'keys', 'values', Symbol.iterator, 'size'],
  },
```
however this requires some fixes and fails the for...of test, so it's not a drop in suggestion

Reply:
fixed